File: 0547_Number of Provinces/NumberofProvinces_UnionFind.cs
public class Solution {
    public int FindCircleNum(int[][] isConnected) {
        if(isConnected == null || isConnected.Length == 0) return 0;
        var n = isConnected.Length;
        var unionFind = new UnionFindSet(n);
        
        for(int u=0; u<n; u++) {
            for(int v=0; v<isConnected[u].Length; v++) {
                if(isConnected[u][v] == 1)
                    unionFind.Union(u,v);
            }
        }
        
        var set = new HashSet<int>();
        for(int u=0;u<n;u++)
            set.Add(unionFind.Find(u));
        
        return set.Count;
    }
}

public class UnionFindSet {
    private int[] parent;
    private int[] rank;
    
    public UnionFindSet(int n) {
        parent = new int[n];
        rank = new int[n];
        
        for(int u=0; u<n; u++)
        {
            parent[u] = u;
            rank[u] = 1;
        }
    }
    
    public int Find(int u) {
        if(u != parent[u])
            return Find(parent[u]);
        return u;
    }
    
    public bool Union(int u, int v) {
        var rootU = Find(u);
        var rootV = Find(v);
        
        if(rootU == rootV) return false;
        
        if(rank[rootU] > rank[rootV])
            parent[rootV] = rootU;
        else if(rank[rootU] < rank[rootV])
            parent[rootU] = rootV;
        else{
            parent[rootU] = rootV;
            rank[rootV]++;
        }
            
        return true;
    }
}

File: 0547_Number of Provinces/NumberofProvinces_UnionFind.java
public class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        UnionFindSet uf = new UnionFindSet(n);
        for(int i=0;i<n;i++) {
            int[] connected = isConnected[i];
            for(int j=0;j<connected.length;j++) {
                if(connected[j] == 1) uf.union(i,j);
            }
        }
        
        int count = 0;
        for(int i=0;i<n;i++)
            if(uf.find(i) == i) count++;
        
        return count;
    }
}

 class UnionFindSet{
    private int[] parents;
    private int[] ranks;

    public UnionFindSet(int count){
        parents = new int[count];
        ranks = new int[count];

        for(int i=0;i<parents.length;i++){
            parents[i] = i;
            ranks[i] = 1;
        }
    }

    //return the root node
    public int find(int n){
        if(n != parents[n])
            return find(parents[n]);
        return n;
    }

    public boolean union(int n1, int n2) {
        int rootN1 = find(n1);
        int rootN2 = find(n2);
        if (rootN1 == rootN2) return false;
   
        if (ranks[rootN1] > ranks[rootN2])
            parents[rootN2] = rootN1;           
        else if (ranks[rootN2] > ranks[rootN1])
            parents[rootN1] = rootN2;
        else {
            parents[rootN1] = rootN2;
            ranks[rootN2] += 1;
        }
   
        return true;
    }
}